### DISCUSSION TITLE
Potential new use case: Use of METplus and python embedding for verification of WRF-LES using observations from the ARL tower array

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#2469] [Use Cases] [answered]
@DanielAdriaansen
@j-opatz
@JohnHalleyGotway
The following is a description of the project which was briefly discussed during the METplus Advanced Training Series Session 2 on 15 NOV 2023. During the training, Ms. Tara Jenson of the METplus Development Team suggested that Mr. John Raby of ARL provide a description of the unique challenges anticipated for the project due to the unique nature of the observations data obtained from an array of twelve instrumented towers located in Southern New Mexico. The following link will connect to the part of the training video where I describe how I have used python embedding in the past for a related project.https://usg01.safelinks.protection.office365.us/?url=https%3A%2F%2Fyoutu.be%2Fv40xA09zwr8%3Ft%3D2692&data=05%7C02%7Cjohn.w.raby2.civ%40army.mil%7C4397b3f8d131498ca95d08dc17a63398%7Cfae6d70f954b481192b60530d6f84c43%7C0%7C0%7C638411249828070192%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=5gmMlZxsXpUOqgWN4ekFqP%2FuzzXtcDqYBTbR57RT2yQ

The project goal is to evaluate simulations produced by the WRF which was run in a VLES/LES configuration. The ground truth data consists of high temporal resolution measurements of air temperature, relative humidity, pressure, and the horizontal wind components produced by the tower mounted sensors at the 2m and 10m AGL levels. The observations data for each tower is contained in text files. The wind component data (m/s) from sonic anemometers collected at 20 Hz is contained in two separate files. One file for the 2m data and one file for the 10m data. The temperature (C), relative humidity (%) and pressure data (hPa) (thermodynamic data) collected at 1 Hz for the 2m and 10m levels are contained in one text file. The format of the observations is non-standard, consisting of one row of data corresponding to each time. No location or elevation data is included with the measurements. This data is available in a separate text file. The following provides examples of the data formats for the wind data and the thermodynamic data.

Wind data: (no headers)
YYYY-MM-DDTHH:MM:SS.SSSSSS -HH:00 (departure from UTC) U V W TMP (C) Sound Speed (m/s)
2022-07-06T00:00:00.040517-07:00 1.55 0.39 0.04 26.53 347.44
2022-07-06T00:00:00.090541-07:00 1.74 0.49 0.04 26.62 347.49
2022-07-06T00:00:00.140514-07:00 1.75 0.72 0.06 26.55 347.45
Note: -6:00 indicates MDT; -7:00 indicates MST
Note: The TMP measurements from the sonic anemometer are not used due to concerns about accuracy.

Thermo data (has col headers):
YYYY-MM-DDTHH:MM:SS -HH:00 (departure from UTC) secCount PRES2m TMP2m RH2m TMP10m
2022-07-06T00:00:01-06:00 1 869.25781 25.83012 43.124381 25.78677
2022-07-06T00:00:02-06:00 2 869.29452 25.83035 43.124432 25.78461
2022-07-06T00:00:03-06:00 3 869.24963 25.82754 43.120213 25.7832
Note: -6:00 indicates MDT; -7:00 indicates MST

The WRF-LES output was post-processed using the UPP. The output is contained in GRIB files valid at 15-min intervals using UTC time. For verification, the goal is to use 15-min averaged observations of each variable which are valid at the same time as the WRF-LES output.

I have done some preliminary testing to see how MET would handle typical tower data which I extracted manually from the files. I added a fictitious time stamp to match the UTC time of the post-processed WRF-LES output. I added representative elevation values in meters to fulfil the requirement for this data in the MET 11-column format. The following is an example of the data I used as input to MET ASCII2NC which was successfully ingested and converted to NetCDF:

ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39 11 869.25781 1351.39 NA 298.97753
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39 52 869.25781 1351.39 NA 43.187570
ADPUPA MSATWR10 20220705_120000 32.58 -106.67 1349.39 11 868.32431 1359.39 NA 298.939
ADPUPA MSATWR10 20220705_120000 32.58 -106.67 1349.39 33 869.25781 1351.39 NA 1.55
ADPUPA MSATWR10 20220705_120000 32.58 -106.67 1349.39 34 869.25781 1351.39 NA 0.39
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39 33 868.32431 1359.39 NA 1.9
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39 34 868.32431 1359.39 NA 1.06

The following shows the NetCDF output from ASCII2NC in human readable format:

#msg_type s_id valid_time  lat  lon  elv  var_name/gc  level  height  qty  value
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39  11 869.2578 1351.39 NA 298.97753906
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39  52 869.2578 1351.39 NA  43.18756866
ADPUPA MSATWR10 20220705_120000 32.58 -106.67 1349.39  11 868.3243 1359.39 NA 298.93899536
ADPUPA MSATWR10 20220705_120000 32.58 -106.67 1349.39  33 869.2578 1351.39 NA   1.54999995
ADPUPA MSATWR10 20220705_120000 32.58 -106.67 1349.39  34 869.2578 1351.39 NA   0.39000000
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39  33 868.3243 1359.39 NA   1.90000000
ADPSFC MSATWR10 20220705_120000 32.58 -106.67 1349.39  34 868.3243 1359.39 NA   1.05999994

The elevation data I used for col 6 is based on the terrain elevation at the tower base. I adjusted the elevation values in col 9 to correspond to the elevation difference from the tower base to the level of the sensor which is either 2m AGL or 10m AGL. I converted the temperature values from C to K. Since I only have the pressure measurement at the 2 m level, I had to extrapolate to compute the pressure at the 10m level using a standard lapse rate. I assigned the message type ADPSFC so surface verification is performed on the standard 2m level variables (TMP, RH) and the standard 10m level wind component variables (UGRD, VGRD). I assigned the ADPUPA message type to these variables that are at non-standard levels which are 2m UGRD and VGRD wind components and 10m TMP so the verification would be done on vertical pressure levels.

I ran Point-Stat using the post-processed WRF-LES output as the forecast input and the NetCDF observation file as the observation input and it successfully ran. The following shows some of the contents of the Matched Pair data text file:

FCST_LEV OBS_VAR OBS_UNITS OBS_LEV  	OBTYPE	    OBS_LVL   	OBS_ELV    	FCST      OBS
Z2	        TMP             K                  Z2       	ADPSFC     NA		1351.39001        293.04538 298.97754
Z2	        RH               %                 Z2      	ADPSFC     NA		1351.39001        82.1006      43.18757
Z10 	        UGRD          m/s              Z10     	ADPSFC     NA		1359.39001       -1.85389      1.9 
Z10          VGRD          m/s              Z10     	ADPSFC     NA		1359.39001       -1.86718      1.06
P869        TMP             K               P 876-862	ADPUPA   868.3242	1359.39001        292.70077  298.939
P869        UGRD           m/s          P 876-862   ADPUPA   869.2578	1351.39001        -1.44576     1.55
P869        VGRD            m/s         P 876-862   ADPUPA   869.2578	1351.39001        -1.4546       0.39

The forecast and observations appear to be matched as needed to compute the error statistics when combined with the same data from the other towers at the same time or with the same tower at different times. There are 7 matched pairs corresponding to each of the 7 observations coming from the tower (the “TOTAL” values in the matched pair data for each observation are all 1).

I would appreciate your feedback on the approach I used to encode the tower data into the 11-col format. I would also appreciate the opportunity to work with you to develop a method to use METplus and Python embedding to automate the reformatting of the tower data into the 11-col format ready for ingest into ASCII2NC and then Point-Stat. I would appreciate any suggestions you have for possibly automating the process of averaging the tower data to produce 15 min averages which are valid at the top of the hour and at 15 min intervals (UTC) to coincide with the valid times of the forecasts.

I can provide additional samples of the tower data and other data on request. I cannot upload the files in the forum or use FTP, so arrangements have been made with Julie Prestopnik to receive the files from an Army fileserver. She will distribute the files as needed within the DTC to support any potential collaboration. Distribution outside of DTC is not authorized without prior coordination with Mr. John Raby.

**Comment by j-opatz** [comment]
Hi @jwraby,

And thank you for this wealth of information. I'm going to tag @georgemccabe who is the METplus Wrappers repository lead for a little direction on this. I know some work is currently being done to ingest WRF output natively, so a WRF-LES file might be a good candidate for that (unless you want the output post-processed by the UPP). What you've laid out here sounds like the start (and then some) of a great use case. It was unclear to me from your layout if you've utilized ASCII2NC's `time_summary` options, which can provide the 15 minute averages you wanted to automate. If we can access any preliminary scripts and data via Julie to provide your requested feedback, we'd be well on our way.

One thing to consider if this goes on to become a use case in the METplus repository is any data necessary to run the use case will need to be made available to the public. We will not publish anything before that time, but METplus use cases need all of the scripts and input data necessary to run them from a general user.

**Reply by georgemccabe** [reply]
The changes I am implementing to support WRF output natively in MET involves reading NetCDF WRF output. It sounds like this data is post-processed by the UPP and can already be read into MET without issue.

**Reply by jwraby** [reply]
Hi @georgemccabe and @j-opatz. Thanks for reviewing the discussion and providing preliminary inputs/clarifications. Before I provide further input to pursue this as a potential use case, I need to get approvals for proceeding given that the information would need to be released to the public.

**Reply by jwraby** [reply]
I have approval for release of examples of the tower observations and post-processed WRF output. I will provide the files to Julie Prestopnik. If it is decided to develop a use case from this project, I will need to get ARL approval prior to public release. Based on the tip provided by @j-opatz above, I was able to use the time_summary option of ASCII2NC to compute 15-min mean values from an input file containing simulated one-min tower TMP observations. Nice feature and potentially useful for this project.

**Comment by jprestop** [comment]
HI @jwraby.  I just wanted to follow up as I have yet not received a DoD SAFE link to pull data from.

**Reply by jwraby** [reply]
Thanks for checking. Sorry about that. I was delayed and could not get to that point yesterday. I will try to upload the data send you the SAFE link today.

**Reply by jprestop** [reply]
@jwraby There is no rush.  In fact, I'm out of the office soon, so I won't be able to grab them until tomorrow.  I just thought I'd check in since I was thinking about it.

**Reply by jwraby** [reply]
No problem. I will try to get most if not all the data and READMEs which support the project uploaded to the SAFE today since I will be off tomorrow through next Mon. Hopefully you will see the SAFE notification in your email when you com in tomorrow.

**Reply by jwraby** [reply]
@jprestop I just uploaded all the data in 3 separate uploads. The one I sent first contains two Word docs which will hopefully explain all the data and details about project. Download these files first as they should help with understanding all the data down to file descriptions. The one I sent after that contains a test case study and results with simulated tower observations in 11-col format, ASCII2NC output, post-processed WRF output, Point-Stat output and other information described in the Word doc "20220705 test case.docx". The last one I sent contains the raw tower observations files and the post-processed WRF output as described in the Word doc "Test case data for 20220706.docx" for use in development work aimed at any assistance you can provide with automating the averaging of the observations into 15-min averages and reformatting the observations into the 11-col format. Let me know if you have any questions or if you need anything else sent. Looking forward to working with you on this project!

**Reply by jprestop** [reply]
@jwraby Thanks!  I was able to receive all three zip files.  

@j-opatz and @georgemccabe, I put these files on seneca at "/d1/projects/METplus/discussions/2469" in subdirectories based on @jwraby's description.  Please let me know if you have any questions or problems accessing these files.

**Comment by DanielAdriaansen** [comment, marked as answer]
We've opened a GitHub issue to track this work in dtcenter/METplus#2487. Please continue any additional conversation there. Thank you!
